File: scala/src/main/scala/Particle.scala
/***
  * Partical state defines the position in chamber and moving direction
  * @param vector
  * @param position
  */
class Particle private (val vector:Int, val position:Int) {
  /**
    * Move a particle and return new particle state
    * @param speed
    * @return a Particle with new position
    */
  def move(speed:Int):Particle=new Particle(vector,position+speed*vector)
}

/**
  * Companion object for Particle
  */
object Particle {
  /**
    * Create Right-moving particle
    * @param position in Chamber
    * @return new instance of Particle
    */
  def R(position:Int):Particle=new Particle(1,position)

  /**
    * Create Left-moving particle
    * @param position in Chamber
    * @return new instance of Particle
    */
  def L(position:Int):Particle=new Particle(-1,position)
}

File: cs/Animation/Particle.cs
namespace Animation
{
    /// <summary>
    /// Immutable Particle state
    /// Represents Particle position in a Chamber and its direction
    /// </summary>
    public class Particle
    {
        /// <summary>
        /// Vector defiles the moving direction o the Particle
        ///  1: Move Right
        /// -1: Move Left
        /// </summary>
        public int Vector { get; }
        /// <summary>
        /// Position in the chamber
        /// </summary>
        public int Position { get; }
        private Particle(int vector,int position)
        {
            Vector = vector;
            Position = position;
        }
        public Particle Move(int speed)
        {
            return new Particle(Vector, Position + speed * Vector);
        }
        /// <summary>
        /// Factory method that produces right-moving particle
        /// </summary>
        /// <param name="position">Position in chamber</param>
        /// <returns></returns>
        public static Particle R(int position) => new Particle(1, position);
        /// <summary>
        /// Factory method that produces left-moving particle
        /// </summary>
        /// <param name="position">Position in CHamber</param>
        /// <returns></returns>
        public static Particle L(int position) => new Particle(-1, position);
    }
}
